darts/models/forecasting/tft_model.py

```diff
@@ -0,0 +1,915 @@
+
+            # this is only to interpret the output
+            static_covariate_var = torch.zeros((past_target.shape[0], 0),
+                                               dtype=past_target.dtype,
+                                               device=past_target.device)
+
+        if future_covariates is None and static_covariates is None:
+            raise NotImplementedError('make zero tensor if future covariates is None')
+
+        static_context_expanded = self.expand_static_context(
+            context=self.static_context_grn(static_embedding),
+            timesteps=timesteps
+        )
+
+        embeddings_varying_encoder = {name: input_vectors_past[name] for name in self.encoder_variables}
+        embeddings_varying_encoder, encoder_sparse_weights = self.encoder_vsn(
+            x=embeddings_varying_encoder,
+            context=static_context_expanded[:, :encoder_length],
+        )
+
+        embeddings_varying_decoder = {name: input_vectors_future[name] for name in self.decoder_variables}
+        embeddings_varying_decoder, decoder_sparse_weights = self.decoder_vsn(
+            x=embeddings_varying_decoder,
+            context=static_context_expanded[:, encoder_length:],
+        )
+
+        # LSTM
+        # calculate initial state
+        input_hidden = self.static_context_initial_hidden_lstm(static_embedding).expand(
+            self.lstm_layers, -1, -1
+        )
+        input_cell = self.static_context_initial_cell_lstm(static_embedding).expand(
+            self.lstm_layers, -1, -1
+        )
+
+        # run local lstm encoder
+        encoder_out, (hidden, cell) = self.lstm_encoder(
+            x=embeddings_varying_encoder,
+            hx=(input_hidden, input_cell),
+            lengths=encoder_lengths,
+            enforce_sorted=False
+        )
+
+        # run local lstm decoder
+        decoder_out, _ = self.lstm_decoder(
+            x=embeddings_varying_decoder,
+            hx=(hidden, cell),
+            lengths=decoder_lengths,
+            enforce_sorted=False,
+        )
+
+        # post lstm GateAddNorm
+        lstm_out_encoder = self.post_lstm_encoder_gan(x=encoder_out, skip=embeddings_varying_encoder)
+        lstm_out_decoder = self.post_lstm_encoder_gan(x=decoder_out, skip=embeddings_varying_decoder)
+
+        lstm_out = torch.cat([lstm_out_encoder, lstm_out_decoder], dim=1)
+
+        # static enrichment
+        static_context_enriched = self.static_context_enrichment(static_embedding)
+        attn_input = self.static_enrichment_grn(
+            x=lstm_out,
+            context=self.expand_static_context(context=static_context_enriched, timesteps=timesteps)
+        )
+
+        # multi-head attention
+        attn_out, attn_out_weights = self.multihead_attn(
+            q=attn_input[:, encoder_length:],  # query only for predictions
+            k=attn_input,
+            v=attn_input,
+            mask=self.get_attention_mask(
```

Comment:
I guess often this mask will be the the same (when operating on the full input+output chunks), so it might make sense to cache the "full" version to avoid re-generating and moving to device it every time.
It's nitpicking though and I doubt it'd have a big performance impact so we can leave it like that for now :)

Reply:
Included it :)

Reply:
Did you push your changes? I don't see it.

Reply:
Now I did